source/common/http/http_server_properties_cache_impl.cc

```diff
@@ -296,6 +297,60 @@ HttpServerPropertiesCacheImpl::getOrCreateHttp3StatusTracker(const Origin& origi
   return *it->second.h3_status_tracker;
 }
 
+void HttpServerPropertiesCacheImpl::markHttp3Broken(const Origin& origin) {
+  getOrCreateHttp3StatusTracker(origin).markHttp3Broken();
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.use_canonical_suffix_for_quic_brokenness")) {
+    maybeSetHttp3BrokenCanonicalOrigin(origin);
+  }
+}
+
+bool HttpServerPropertiesCacheImpl::isHttp3Broken(const Origin& origin) {
+  absl::optional<Origin> canonical = getHttp3BrokenCanonicalOrigin(origin.hostname_);
+
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.use_canonical_suffix_for_quic_brokenness") ||
+      !canonical.has_value()) {
+    return getOrCreateHttp3StatusTracker(origin).isHttp3Broken();
+  }
+
+  // Note that we don't create a new tracker for the origin if there's already a corresponding one
+  // for the canonical origin.
+  if (auto entry_it = protocols_.find(origin);
+      entry_it != protocols_.end() && entry_it->second.h3_status_tracker != nullptr) {
+    return entry_it->second.h3_status_tracker->isHttp3Broken();
+  }
+
+  if (auto entry_it = protocols_.find(*canonical);
+      entry_it != protocols_.end() && entry_it->second.h3_status_tracker != nullptr) {
```

Comment:
If we have an initial quic hint, it will be in `protocol_`, but the entry won't have a h3 tracker until either markH3Confirmed() or markH3Broken() is called.

Reply:
After using a different map canonical_h3_broken_map_ to track canonical origin for H3 brokenness, if getHttp3BrokenCanonicalOrigin() doesn't return nullopt, the canonical origin must have a valid H3 status, right?  IIUC, the new map is only populated when we markH3Confirmed() or markH3Broken().

Reply:
You are right. If an origin exists in the canonical_h3_broken_map_, it must have a http3 tracker. What would you suggest me to do here? Just use the tracker without checking for nullptr?